### PR TITLE
case Dev 471: HMD default audio behavior 

### PR DESCRIPTION
**Pull request by amerhifi**
https://highfidelity.atlassian.net/browse/DEV-471
https://highfidelity.atlassian.net/browse/DEV-2266
https://highfidelity.atlassian.net/browse/DEV-475
https://highfidelity.atlassian.net/browse/DEV-472

Hmd default device.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Can someone from High Fidelity approve this PR?

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22036/HighFidelity-Beta-PR16381-421cff1.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22036/HighFidelity-Quest-Beta-PR16381-421cff1.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22036/HighFidelity-Beta-PR16381-421cff1.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22036/HighFidelity-Beta-Interface-PR16381-421cff1.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22039/HighFidelity-Beta-PR16381-ec7c8ce.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22039/HighFidelity-Beta-Interface-PR16381-ec7c8ce.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22039/HighFidelity-Beta-PR16381-ec7c8ce.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22039/HighFidelity-Quest-Beta-PR16381-ec7c8ce.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22039/HighFidelity-Beta-PR16381-ec7c8ce.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22039/HighFidelity-Beta-Interface-PR16381-ec7c8ce.exe

**Comment by JeffClinton** [comment]
Pre-merge testing was done earlier today, and an issue was noticed where radio buttons were missing for the desktop settings after launching into HMD.  This is being looked into.  Other than that, things are looking good.  Audio correctly defaults to Rift when launching into simplified UI.

audio - desktop

File: libraries/audio-client/CMakeLists.txt [hunk 1]
```diff
@@ -6,7 +6,7 @@ setup_hifi_library(Network Multimedia ${PLATFORM_QT_COMPONENTS})
 link_hifi_libraries(audio plugins)
 include_hifi_library_headers(shared)
 include_hifi_library_headers(networking)
-
+include_hifi_library_headers(gpu)
```
Comment (kencooke) [inline libraries/audio-client/CMakeLists.txt]: You lost a newline here

File: libraries/audio-client/src/AudioClient.cpp [hunk 2]
```diff
@@ -47,7 +47,7 @@
 #include <SettingHandle.h>
 #include <SharedUtil.h>
 #include <Transform.h>
-
+#include <plugins/DisplayPlugin.h>
```
Comment (kencooke) [inline libraries/audio-client/src/AudioClient.cpp]: Lost a newline here

File: libraries/audio-client/src/AudioClient.cpp [hunk 3]
```diff
@@ -81,20 +81,48 @@ Mutex _recordMutex;
 
 HifiAudioDeviceInfo defaultAudioDeviceForMode(QAudio::Mode mode);
 
+#include <shared/QtHelpers.h>
```
Comment (kencooke) [inline libraries/audio-client/src/AudioClient.cpp]: Should be with other related includes

File: libraries/audio-client/src/AudioClient.cpp [hunk 4]
```diff
@@ -558,8 +594,8 @@ HifiAudioDeviceInfo defaultAudioDeviceForMode(QAudio::Mode mode) {
     }
 #endif
     // fallback for failed lookup is the default device
-    return (mode == QAudio::AudioInput) ? HifiAudioDeviceInfo(QAudioDeviceInfo::defaultInputDevice(), true,mode) : 
-        HifiAudioDeviceInfo(QAudioDeviceInfo::defaultOutputDevice(), true, mode);
+    return (mode == QAudio::AudioInput) ? HifiAudioDeviceInfo(QAudioDeviceInfo::defaultInputDevice(), true,mode, HifiAudioDeviceInfo::desktop) :
```
Comment (kencooke) [inline libraries/audio-client/src/AudioClient.cpp]: formatting should be `true, mode`

File: libraries/audio-client/src/AudioClient.cpp [hunk 5]
```diff
@@ -532,7 +568,7 @@ HifiAudioDeviceInfo defaultAudioDeviceForMode(QAudio::Mode mode) {
     HifiAudioDeviceInfo foundDevice;
     foreach(QAudioDeviceInfo audioDevice, devices) {
         if (audioDevice.deviceName().trimmed() == deviceName.trimmed()) {
-            foundDevice=HifiAudioDeviceInfo(audioDevice,true,mode);
+            foundDevice=HifiAudioDeviceInfo(audioDevice,true,mode, HifiAudioDeviceInfo::desktop);
```
Comment (kencooke) [inline libraries/audio-client/src/AudioClient.cpp]: formatting should be `foundDevice = HifiAudioDeviceInfo(audioDevice, true, mode, `

File: libraries/audio-client/src/HifiAudioDeviceInfo.h [hunk 6]
```diff
@@ -23,19 +23,28 @@ class HifiAudioDeviceInfo : public QObject {
...
     HifiAudioDeviceInfo() : QObject() {}
     HifiAudioDeviceInfo(const HifiAudioDeviceInfo &deviceInfo) : QObject(){
         _audioDeviceInfo = deviceInfo.getDevice();
         _mode = deviceInfo.getMode();
         _isDefault = deviceInfo.isDefault();
+        _deviceType = deviceInfo.getDeviceType();
     }
 
-    HifiAudioDeviceInfo(QAudioDeviceInfo deviceInfo, bool isDefault, QAudio::Mode mode) :
+
```
Comment (kencooke) [inline libraries/audio-client/src/HifiAudioDeviceInfo.h]: Extra newline here

File: libraries/audio-client/src/HifiAudioDeviceInfo.h [hunk 7]
```diff
@@ -52,15 +61,17 @@ class HifiAudioDeviceInfo : public QObject {
     QAudioDeviceInfo getDevice() const { return _audioDeviceInfo; }
     bool isDefault() const { return _isDefault; }
     QAudio::Mode getMode() const { return _mode; }
-
+    DeviceType getDeviceType() const { return _deviceType; }
     HifiAudioDeviceInfo& operator=(const HifiAudioDeviceInfo& other);
     bool operator==(const HifiAudioDeviceInfo& rhs) const;
     bool operator!=(const HifiAudioDeviceInfo& rhs) const;
 
+
```
Comment (kencooke) [inline libraries/audio-client/src/HifiAudioDeviceInfo.h]: Extra newline here

File: libraries/plugins/src/plugins/PluginManager.cpp [hunk 8]
```diff
@@ -221,7 +221,21 @@ const OculusPlatformPluginPointer PluginManager::getOculusPlatformPlugin() {
     return oculusPlatformPlugin;
 }
 
-const DisplayPluginList& PluginManager::getDisplayPlugins() {
+#include <shared/QtHelpers.h>
```
Comment (kencooke) [inline libraries/plugins/src/plugins/PluginManager.cpp]: Should be with related includes above

File: libraries/plugins/src/plugins/PluginManager.cpp [hunk 9]
```diff
@@ -221,7 +221,21 @@ const OculusPlatformPluginPointer PluginManager::getOculusPlatformPlugin() {
...
+#include <QThreadPool>
+
+
```
Comment (kencooke) [inline libraries/plugins/src/plugins/PluginManager.cpp]: Extra newline here

File: libraries/plugins/src/plugins/PluginManager.h [hunk 10]
```diff
@@ -51,7 +51,7 @@ class PluginManager : public QObject, public Dependency {
 
     using PluginFilter = std::function<bool(const QJsonObject&)>;
     void setPluginFilter(PluginFilter pluginFilter) { _pluginFilter = pluginFilter; }
-
+    Q_INVOKABLE DisplayPluginList getAllDisplayPlugins();
```
Comment (kencooke) [inline libraries/plugins/src/plugins/PluginManager.h]: Lost a newline here

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22050/HighFidelity-Beta-PR16381-68681fd.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22050/HighFidelity-Beta-Interface-PR16381-68681fd.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22050/HighFidelity-Beta-PR16381-68681fd.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22050/HighFidelity-Quest-Beta-PR16381-68681fd.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22050/HighFidelity-Beta-PR16381-68681fd.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22050/HighFidelity-Beta-Interface-PR16381-68681fd.exe

File: libraries/audio-client/src/AudioClient.cpp [hunk 11]
```diff
@@ -568,7 +570,7 @@ HifiAudioDeviceInfo defaultAudioDeviceForMode(QAudio::Mode mode) {
     HifiAudioDeviceInfo foundDevice;
     foreach(QAudioDeviceInfo audioDevice, devices) {
         if (audioDevice.deviceName().trimmed() == deviceName.trimmed()) {
-            foundDevice=HifiAudioDeviceInfo(audioDevice,true,mode, HifiAudioDeviceInfo::desktop);
+            foundDevice=HifiAudioDeviceInfo(audioDevice, true, mode, HifiAudioDeviceInfo::desktop);
```
Comment (kencooke) [inline libraries/audio-client/src/AudioClient.cpp]: spaces around the `=`

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22053/HighFidelity-Beta-PR16381-a860219.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22053/HighFidelity-Beta-Interface-PR16381-a860219.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22053/HighFidelity-Beta-PR16381-a860219.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22053/HighFidelity-Quest-Beta-PR16381-a860219.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22053/HighFidelity-Beta-PR16381-a860219.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22053/HighFidelity-Beta-Interface-PR16381-a860219.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22056/HighFidelity-Beta-PR16381-0410a45.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22056/HighFidelity-Beta-Interface-PR16381-0410a45.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22056/HighFidelity-Beta-PR16381-0410a45.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22056/HighFidelity-Beta-Interface-PR16381-0410a45.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22056/HighFidelity-Beta-PR16381-0410a45.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22056/HighFidelity-Quest-Beta-PR16381-0410a45.apk)

**Comment by JeffClinton** [comment]
Pre-merge testing

**Comment by JeffClinton** [comment]
Audio is no longer defaulting to HMD on the VR tab when launching into simplified UI.  There also appears to be quite a bit of audioclient spam in the logs.

Steps to repro:
1. Clean uninstall any existing builds
2. Install PR client to PC with HMD but do not launch
3. Open command prompt
4. Launch into simplified UI using this command (edit as necessary)
"C:\Program Files\High Fidelity Interface - PR16381\interface.exe" --defaultScriptsOverride "C:/Program Files/High Fidelity Interface - PR16381/scripts/simplifiedUIBootstrapper.js"


audio - vr build 2
[Logs.zip](https://github.com/highfidelity/hifi/files/3757925/Logs.zip)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22077/HighFidelity-Beta-PR16381-179f5e0.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22077/HighFidelity-Beta-Interface-PR16381-179f5e0.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22077/HighFidelity-Beta-PR16381-179f5e0.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22077/HighFidelity-Beta-Interface-PR16381-179f5e0.exe

**Comment by JeffClinton** [comment]
Pre-merge testing

File: libraries/audio-client/src/AudioClient.cpp [hunk 12]
```diff
@@ -1816,8 +1816,8 @@ void AudioClient::outputFormatChanged() {
 bool AudioClient::switchInputToAudioDevice(const HifiAudioDeviceInfo inputDeviceInfo, bool isShutdownRequest) {
     Q_ASSERT_X(QThread::currentThread() == thread(), Q_FUNC_INFO, "Function invoked on wrong thread");
 
-    qCDebug(audioclient) << __FUNCTION__ << "inputDeviceInfo: [" << _inputDeviceInfo.deviceName() << ":" << _inputDeviceInfo.getDevice().deviceName() 
-        <<"-- new device: "<< inputDeviceInfo.deviceName() <<":"<< inputDeviceInfo.getDevice().deviceName() << "]";
+    qCDebug(audioclient) << __FUNCTION__ << "_inputDeviceInfo: [" << _inputDeviceInfo.deviceName() << ":" << _inputDeviceInfo.getDevice().deviceName() 
+        << "-- inputDeviceInfo: " << inputDeviceInfo.deviceName() << ":" << inputDeviceInfo.getDevice().deviceName() << "]";
```
Comment (kencooke) [inline libraries/audio-client/src/AudioClient.cpp]: qDebug() already pads with spaces, so you don't want leading/trailing spaces in your strings here

File: libraries/audio-client/src/AudioClient.cpp [hunk 13]
```diff
@@ -1869,8 +1869,9 @@ bool AudioClient::switchInputToAudioDevice(const HifiAudioDeviceInfo inputDevice
     }
 
     if (!inputDeviceInfo.getDevice().isNull()) {
-        qCDebug(audioclient) << "The audio input device " << inputDeviceInfo.deviceName() << "is available.";
+        qCDebug(audioclient) << "The audio input device " << inputDeviceInfo.deviceName() << ":" << inputDeviceInfo.getDevice().deviceName() << " is available.";
```
Comment (kencooke) [inline libraries/audio-client/src/AudioClient.cpp]: extra spaces

File: libraries/audio-client/src/AudioClient.cpp [hunk 14]
```diff
@@ -2053,8 +2054,8 @@ void AudioClient::noteAwakening() {
 bool AudioClient::switchOutputToAudioDevice(const HifiAudioDeviceInfo outputDeviceInfo, bool isShutdownRequest) {
     Q_ASSERT_X(QThread::currentThread() == thread(), Q_FUNC_INFO, "Function invoked on wrong thread");
     
-    qCDebug(audioclient) << __FUNCTION__ << "inputDeviceInfo: [" << _outputDeviceInfo.deviceName() << ":" << _outputDeviceInfo.getDevice().deviceName()
-        << "-- new device: " << outputDeviceInfo.deviceName() << ":" << outputDeviceInfo.getDevice().deviceName() << "]";
+    qCDebug(audioclient) << __FUNCTION__ << "_outputdeviceInfo: [" << _outputDeviceInfo.deviceName() << ":" << _outputDeviceInfo.getDevice().deviceName()
+        << "-- outputDeviceInfo: " << outputDeviceInfo.deviceName() << ":" << outputDeviceInfo.getDevice().deviceName() << "]";
```
Comment (kencooke) [inline libraries/audio-client/src/AudioClient.cpp]: extra spaces

File: libraries/audio-client/src/AudioClient.cpp [hunk 15]
```diff
@@ -2110,7 +2111,9 @@ bool AudioClient::switchOutputToAudioDevice(const HifiAudioDeviceInfo outputDevi
     }
 
     if (!outputDeviceInfo.getDevice().isNull()) {
-        qCDebug(audioclient) << "The audio output device " << outputDeviceInfo.deviceName() << "is available.";
+        qCDebug(audioclient) << "The audio output device " << outputDeviceInfo.deviceName() << ":" << outputDeviceInfo.getDevice().deviceName() << " is available.";
```
Comment (kencooke) [inline libraries/audio-client/src/AudioClient.cpp]: extra spaces

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22089/HighFidelity-Beta-PR16381-ac8b664.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22089/HighFidelity-Quest-Beta-PR16381-ac8b664.apk)

**Comment by JeffClinton** [comment]
Pre-merge testing paused due to a shutdown issue

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22089/HighFidelity-Beta-PR16381-ac8b664.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22089/HighFidelity-Beta-Interface-PR16381-ac8b664.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22116/HighFidelity-Beta-PR16381-5dd5447.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/22116/HighFidelity-Quest-Beta-PR16381-5dd5447.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22116/HighFidelity-Beta-PR16381-5dd5447.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/22116/HighFidelity-Beta-Interface-PR16381-5dd5447.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22116/HighFidelity-Beta-PR16381-5dd5447.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/22116/HighFidelity-Beta-Interface-PR16381-5dd5447.exe

**Comment by JeffClinton** [comment]
Pre-merge testing

**Comment by JeffClinton** [comment]
Pre-merge test PASS
- Desktop and HMD default to correct settings
- No longer experiencing shut down hang

**Pull request merged**